src/bdc/steps/analyze_reviews.py

```diff
@@ -0,0 +1,216 @@
+            sentiment_score = response.choices[0].message.content
+            log.debug(f"GPT response {sentiment_score}")
+            if sentiment_score and sentiment_score != self.no_answer:
+                return float(sentiment_score)
+            else:
+                log.info("No valid sentiment score found in the response.")
+                return None
+
+        except (
+            openai.APITimeoutError,
+            openai.APIConnectionError,
+            openai.BadRequestError,
+            openai.AuthenticationError,
+            openai.PermissionDeniedError,
+        ) as e:
+            log.error(f"An error occurred with GPT API: {e}")
+        except Exception as e:
+            log.error(f"An unexpected error occurred: {e}")
+
+        # Return None if any exception occurred
+        return None
+
+    def extract_text_from_reviews(self, reviews_list):
+        """
+        extracts text from reviews and removes line characters.
+        """
+        reviews_texts = [review.get("text", None) for review in reviews_list]
+        review_texts_formatted = [
+            review.strip().replace("\n", " ") for review in reviews_texts if review
+        ]
+        return review_texts_formatted
+
+    def num_tokens_from_string(self, text: str):
+        """Returns the number of tokens in a text string."""
+        encoding = tiktoken.get_encoding(self.model_encoding_name)
+        num_tokens = len(encoding.encode(text))
+        return num_tokens
+
+    def batch_reviews(self, reviews, max_tokens=4096):
+        """
+        Batchs reviews into batches so that every batch has the size less than max_tokens
+        """
+        batches = []
+        current_batch = []
+        current_count = self.num_tokens_from_string(
+            self.user_message_for_sentiment_analysis
+        )
+
+        for review in reviews:
+            token_count = self.num_tokens_from_string(review)
+            if current_count + token_count > max_tokens:
+                batches.append(current_batch)
+                current_batch = [review]
+                current_count = token_count
+            else:
+                current_batch.append(review)
+                current_count += token_count
+
+        if current_batch:
+            batches.append(current_batch)
+
+        return batches
+
+    def fetch_reviews(self, reviews_path):
+        try:
+            # reviews_path always starts with "./data/.."
+            full_path = os.path.abspath(
+                os.path.join(os.path.dirname(__file__), "../../" + reviews_path)
+            )
+            with open(full_path, "r", encoding="utf-8") as reviews_json:
+                reviews = json.load(reviews_json)
+                return reviews
+        except:
+            log.warning(f"Error loading reviews from path {full_path}.")
+            # Return empty list if any exception occurred or status is not OK
+            return []
+
+    def fetch_website(self, place_id):
```

Comment:
Is this currently used? Please access the website directly from the dataframe in the future to avoid redundant API calls.

Reply:
deleted.